package/collection2/collection2.js

```diff
@@ -135,261 +139,344 @@ Mongo.Collection.prototype.attachSchema = function c2AttachSchema(ss, options) {
-          target = query[selector];
+          target = query[selector]
         }
 
         // we need to compare given selector with doc property or option to
-        // find right schema
+        //  find the right schema
         if (target !== undefined && target === schema.selector[selector]) {
-          return schema.schema;
+          return schema.schema
         }
       }
       if (schemas[0]) {
-        return schemas[0].schema;
+        return schemas[0].schema
       } else {
-        throw new Error("No default schema");
+        throw new Error('No default schema')
       }
     }
 
-    return null;
-  };
-});
+    return null
+  }
+})
+
+function _methodMutation (async, methodName) {
+  const _super = Meteor.isFibersDisabled
+    ? Mongo.Collection.prototype[methodName]
+    : Mongo.Collection.prototype[methodName.replace('Async', '')]
 
-// Wrap DB write operation methods
-['insert', 'update'].forEach((methodName) => {
-  const _super = Mongo.Collection.prototype[methodName];
-  Mongo.Collection.prototype[methodName] = function(...args) {
-    let options = (methodName === "insert") ? args[1] : args[2];
+  if (!_super) return
+
+  Mongo.Collection.prototype[methodName] = function (...args) {
+    let options = (isInsertType(methodName)) ? args[1] : args[2]
 
     // Support missing options arg
-    if (!options || typeof options === "function") {
-      options = {};
+    if (!options || typeof options === 'function') {
+      options = {}
     }
 
+    let validationContext = {}
+    let error
     if (this._c2 && options.bypassCollection2 !== true) {
-      let userId = null;
+      let userId = null
       try { // https://github.com/aldeed/meteor-collection2/issues/175
-        userId = Meteor.userId();
-      } catch (err) {}
+        userId = Meteor.userId()
+      } catch (err) {
+      }
 
-      args = doValidate(
+      [args, validationContext] = doValidate(
         this,
         methodName,
         args,
         Meteor.isServer || this._connection === null, // getAutoValues
         userId,
-        Meteor.isServer // isFromTrustedCode
-      );
+        Meteor.isServer, // isFromTrustedCode
+        async
+      )
+
       if (!args) {
-        // doValidate already called the callback or threw the error so we're done.
+        // doValidate already called the callback or threw the error, so we're done.
         // But insert should always return an ID to match core behavior.
-        return methodName === "insert" ? this._makeNewID() : undefined;
+        return isInsertType(methodName) ? this._makeNewID() : undefined
       }
     } else {
       // We still need to adjust args because insert does not take options
-      if (methodName === "insert" && typeof args[1] !== 'function') args.splice(1, 1);
+      if (isInsertType(methodName) && typeof args[1] !== 'function') args.splice(1, 1)
+    }
+
+    if (async && !Meteor.isFibersDisabled) {
```

Comment:
Sorry for late review here. @klablink could you please add one or two lines of comments and summarize what's happening here in this if block.

I also  would like to know about the logic behind the if-clause, because `Meteor.isAsyncDisabled` might be gone at some day, right?